src/lmql/runtime/bopenai/openai_api.py

```diff
@@ -58,16 +59,62 @@ async def complete(**kwargs):
 global tokenizer
 tokenizer = None
 
+
+class TiktokenTokenizer:
+    def __init__(self):
+        self.enc = tiktoken.get_encoding("cl100k_base")
+        self.vocab = {self.enc.decode([i]): i for i in range(100256)}
+        for i in range(100256, self.enc.max_token_value):
+            self.vocab[f"<|special_{i}|>"] = i
+
+    def encode(self, text):
+        return self.enc.encode(text)
+
+    def tokenize(self, text):
+        return [self.enc.decode([i]) for i in self.enc.encode(text)]
+
+    def decode(self, ids):
+        return self.enc.decode(ids)
+
+    def __call__(self, text_or_list):
+        if isinstance(text_or_list, str):
+            input_ids = self.encode(text_or_list)
+        else:
+            input_ids = [self.encode(text) for text in text_or_list]
+
+        return {
+            "input_ids": input_ids,
+        }
+
+    @property
+    def vocab_size(self):
+        return self.enc.max_token_value
+    
+    @property
+    def eos_token_id(self):
+        return self.enc.eot_token
+
+    @property
+    def bos_token_id(self):
+        return self.enc.encode_single_token("<|endofprompt|>")
```

Comment:
There is no `<BOS>` token in `cl100k_base` - from the brief look, it didn't seem like we're actually sending this to the API?